curvefs/src/metaserver/s3compact_wq_impl.cpp

```diff
@@ -0,0 +1,418 @@
+/*
+ *  Copyright (c) 2021 NetEase Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/*
+ * @Project: curve
+ * @Date: 2021-09-07
+ * @Author: majie1
+ */
+
+#include "curvefs/src/metaserver/s3compact_wq_impl.h"
+
+#include <algorithm>
+#include <deque>
+#include <list>
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+using ::curve::common::Configuration;
+using ::curve::common::InitS3AdaptorOption;
+using ::curve::common::S3Adapter;
+using ::curve::common::S3AdapterOption;
+using ::curve::common::TaskThreadPool;
+
+namespace curvefs {
+namespace metaserver {
+
+void S3CompactWorkQueueImpl::Enqueue(std::shared_ptr<InodeStorage> inodeStorage,
+                                     InodeKey inodeKey, PartitionInfo pinfo,
+                                     CopysetNode* copysetNode) {
+    std::unique_lock<std::mutex> guard(mutex_);
+
+    // inodeKey already in working queue, just return
+    if (std::find(compactingInodes_.begin(), compactingInodes_.end(),
+                  inodeKey) != compactingInodes_.end()) {
+        return;
+    }
+
+    while (IsFullUnlock()) {
+        notFull_.wait(guard);
+    }
+    compactingInodes_.push_back(inodeKey);
+    auto task = std::bind(&S3CompactWorkQueueImpl::CompactChunks, this,
+                          inodeStorage, inodeKey, pinfo, copysetNode);
+    queue_.push_back(std::move(task));
+    notEmpty_.notify_one();
+}
+
+std::function<void()> S3CompactWorkQueueImpl::Dequeue() {
+    std::unique_lock<std::mutex> guard(mutex_);
+    while (queue_.empty() && running_.load(std::memory_order_acquire)) {
+        notEmpty_.wait(guard);
+    }
+    std::function<void()> task;
+    if (!queue_.empty()) {
+        task = std::move(queue_.front());
+        queue_.pop_front();
+        compactingInodes_.pop_front();
+        notFull_.notify_one();
+    }
+
+    return task;
+}
+
+void S3CompactWorkQueueImpl::ThreadFunc() {
+    while (running_.load(std::memory_order_acquire)) {
+        Task task(Dequeue());
+        if (task) {
+            task();
+        }
+    }
+}
+
+std::vector<uint64_t> S3CompactWorkQueueImpl::getNeedCompact(
+    const ::google::protobuf::Map<uint64_t, S3ChunkInfoList>& s3chunkinfoMap) {
+    std::vector<uint64_t> needCompact;
+    int maxChunkPerIndex;
```

Comment:
recommend to make these two vars as member function, and load option when init or run.

Reply:
fixed.